src/vcf/validator.cpp

```diff
@@ -122,13 +126,63 @@ namespace ebi
     }
 
     bool is_valid_vcf_file(std::istream &input,
-                           ebi::vcf::Parser &validator,
+                           const std::string &sourceName,
+                           ValidationLevel validationLevel,
+                           Ploidy ploidy,
                            std::vector<std::unique_ptr<ebi::vcf::ReportWriter>> &outputs)
     {
+        std::vector<char> line;
+        ebi::util::readline(input, line);
+        ebi::vcf::Version version;
+        try {
+            version = detect_version(line);
+        } catch (FileformatError * error) {
+            for (auto &output : outputs) {
+                output->write_error(*error);
+            }
+            return false;
+        }
+        std::unique_ptr<Parser> validator = build_parser(sourceName, validationLevel, version, ploidy);
+        return validate(line, input, *validator, outputs);
+    }
 
+    Version detect_version(const std::vector<char> &line)
```

Comment:
What about something like this, using [std::basic_string::substr](http://en.cppreference.com/w/cpp/string/basic_string/substr)? Not sure if the indexes are right, please check.

```c++
std::string str(line.begin(), line.end());

if (str.substr(0, 13) == "##fileformat=") {
    if (str.substr(14) == "VCFv4.1") {
        return Version::v41;
    } else if (str.substr(14) == "VCFv4.2") {
        return Version::v42;
    } else if (str.substr(14) == "VCFv4.3") {
        return Version::v43;
    }
}

throw new FileformatError{1, "The fileformat declaration is not valid (the line must start with "
    + common_substring + " and the value must be one of 'VCFv4.1', 'VCFv4.2' or 'VCFv4.3')"};
}
```